File: suc/kotlin/048.kt

package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.BottomAppBar
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Shapes
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.Typography
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat

private const val NAME = "063*LiveShop*en"
private const val UI_TYPE = "ECommerce"
private const val STYLE_THEME = "Modern Blue"
private const val LANG = "en"
private const val BASELINE_SIZE = "720x1280"

object AppTokens {
    object Colors {
        val primary = Color(0xFF1E3A8A)
        val secondary = Color(0xFF3B82F6)
        val tertiary = Color(0xFF60A5FA)
        val background = Color(0xFFF5F8FF)
        val surface = Color(0xFFFFFFFF)
        val surfaceVariant = Color(0xFFE8EEFA)
        val outline = Color(0xFFCBD5E1)
        val success = Color(0xFF22C55E)
        val warning = Color(0xFFF59E0B)
        val error = Color(0xFFEF4444)
        val onPrimary = Color(0xFFFFFFFF)
        val onSecondary = Color(0xFFFFFFFF)
        val onTertiary = Color(0xFF0B1220)
        val onBackground = Color(0xFF0B1220)
        val onSurface = Color(0xFF0B1220)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 28.sp, fontWeight = FontWeight.Bold)
        val title = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.Medium)
        val body = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.Normal)
        val label = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.Medium)
    }
    object Shapes {
        val small = RoundedCornerShape(6.dp)
        val medium = RoundedCornerShape(10.dp)
        val large = RoundedCornerShape(16.dp)
    }
    object Spacing {
        val sm = 6.dp
        val md = 10.dp
        val lg = 16.dp
        val xl = 24.dp
        val xxl = 36.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(2.dp, 4.dp, 2.dp, 0.12f)
        val level2 = ShadowSpec(6.dp, 8.dp, 4.dp, 0.15f)
        val level3 = ShadowSpec(10.dp, 12.dp, 6.dp, 0.18f)
    }
}

private val AppColorScheme = lightColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    onSecondary = AppTokens.Colors.onSecondary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onBackground,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface,
    surfaceVariant = AppTokens.Colors.surfaceVariant,
    outline = AppTokens.Colors.outline
)

private val AppTypography = Typography(
    displayLarge = AppTokens.TypographyTokens.display,
    titleMedium = AppTokens.TypographyTokens.title,
    bodyMedium = AppTokens.TypographyTokens.body,
    labelMedium = AppTokens.TypographyTokens.label
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        shapes = Shapes(
            small = AppTokens.Shapes.small,
            medium = AppTokens.Shapes.medium,
            large = AppTokens.Shapes.large
        ),
        content = content
    )
}

data class LiveItem(val id: Int, val title: String, val viewers: Int, val price: String)

@Composable
fun RootScreen() {
    val liveList = remember {
        listOf(
            LiveItem(1, "Smartphone Flash Deal", 1234, "$699"),
            LiveItem(2, "Gaming Chair Special", 880, "$249"),
            LiveItem(3, "Headphones Clearance", 1640, "$99"),
            LiveItem(4, "Mechanical Keyboard", 512, "$129")
        )
    }
    Scaffold(
        contentWindowInsets = androidx.compose.foundation.layout.WindowInsets(0),
        floatingActionButton = {
            FloatingActionButton(
                onClick = {},
                containerColor = AppTokens.Colors.secondary,
                contentColor = AppTokens.Colors.onSecondary,
                shape = CircleShape
            ) {
                Text("+", style = AppTokens.TypographyTokens.title)
            }
        },
        bottomBar = {
            BottomAppBar(
                containerColor = AppTokens.Colors.surface,
                contentColor = AppTokens.Colors.onSurface
            ) {
                Row(
                    modifier = Modifier.fillMaxWidth().padding(horizontal = AppTokens.Spacing.lg),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text("Home", color = AppTokens.Colors.primary, style = AppTokens.TypographyTokens.body)
                    Text("Live", color = AppTokens.Colors.onSurface, style = AppTokens.TypographyTokens.body)
                    Text("Cart", color = AppTokens.Colors.onSurface, style = AppTokens.TypographyTokens.body)
                    Text("Profile", color = AppTokens.Colors.onSurface, style = AppTokens.TypographyTokens.body)
                }
            }
        },
        containerColor = AppTokens.Colors.background
    ) { pad ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(pad)
                .background(
                    Brush.verticalGradient(
                        listOf(
                            AppTokens.Colors.background,
                            AppTokens.Colors.surfaceVariant,
                            AppTokens.Colors.background
                        )
                    )
                )
                .padding(AppTokens.Spacing.lg),
            verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md)
        ) {
            Text("Live Shop Events", style = AppTokens.TypographyTokens.display, color = AppTokens.Colors.primary)
            LazyColumn(
                modifier = Modifier.fillMaxSize(),
                verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md),
                contentPadding = PaddingValues(bottom = 96.dp)
            ) {
                items(liveList) { item ->
                    Card(
                        shape = AppTokens.Shapes.large,
                        colors = CardDefaults.cardColors(containerColor = AppTokens.Colors.surface),
                        elevation = CardDefaults.cardElevation(AppTokens.ElevationMapping.level1.elevation)
                    ) {
                        Column(
                            modifier = Modifier
                                .fillMaxWidth()
                                .padding(AppTokens.Spacing.lg),
                            verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.sm)
                        ) {
                            Box(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .height(160.dp)
                                    .clip(AppTokens.Shapes.medium)
                                    .background(AppTokens.Colors.secondary)
                            )
                            Text(item.title, style = AppTokens.TypographyTokens.title, color = AppTokens.Colors.onSurface)
                            Text("${item.viewers} viewers", style = AppTokens.TypographyTokens.body, color = AppTokens.Colors.tertiary)
                            Row(
                                modifier = Modifier.fillMaxWidth(),
                                horizontalArrangement = Arrangement.SpaceBetween,
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                Text(item.price, style = AppTokens.TypographyTokens.title, color = AppTokens.Colors.primary)
                                Button(
                                    onClick = {},
                                    colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.primary, contentColor = AppTokens.Colors.onPrimary),
                                    shape = AppTokens.Shapes.medium
                                ) {
                                    Text("Join Live", style = AppTokens.TypographyTokens.label)
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.systemBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.systemBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFF5F8FF)
@Composable
fun PreviewRoot() {
    AppTheme {
        RootScreen()
    }
}

File: suc/swift/048.swift

import SwiftUI

// MARK: - AppTokens (Equivalent to Kotlin's AppTokens object)
// 包含应用的所有设计系统常量，便于统一管理和修改
struct AppTokens {
    struct Colors {
        static let primary = Color(red: 0x1E / 255.0, green: 0x3A / 255.0, blue: 0x8A / 255.0)
        static let secondary = Color(red: 0x3B / 255.0, green: 0x82 / 255.0, blue: 0xF6 / 255.0)
        static let tertiary = Color(red: 0x60 / 255.0, green: 0xA5 / 255.0, blue: 0xFA / 255.0)
        static let background = Color(red: 0xF5 / 255.0, green: 0xF8 / 255.0, blue: 0xFF / 255.0)
        static let surface = Color(red: 0xFF / 255.0, green: 0xFF / 255.0, blue: 0xFF / 255.0)
        static let surfaceVariant = Color(red: 0xE8 / 255.0, green: 0xEE / 255.0, blue: 0xFA / 255.0)
        static let outline = Color(red: 0xCB / 255.0, green: 0xD5 / 255.0, blue: 0xE1 / 255.0)
        static let success = Color(red: 0x22 / 255.0, green: 0xC5 / 255.0, blue: 0x5E / 255.0)
        static let warning = Color(red: 0xF5 / 255.0, green: 0x9E / 255.0, blue: 0x0B / 255.0)
        static let error = Color(red: 0xEF / 255.0, green: 0x44 / 255.0, blue: 0x44 / 255.0)
        static let onPrimary = Color(red: 0xFF / 255.0, green: 0xFF / 255.0, blue: 0xFF / 255.0)
        static let onSecondary = Color(red: 0xFF / 255.0, green: 0xFF / 255.0, blue: 0xFF / 255.0)
        static let onTertiary = Color(red: 0x0B / 255.0, green: 0x12 / 255.0, blue: 0x20 / 255.0)
        static let onBackground = Color(red: 0x0B / 255.0, green: 0x12 / 255.0, blue: 0x20 / 255.0)
        static let onSurface = Color(red: 0x0B / 255.0, green: 0x12 / 255.0, blue: 0x20 / 255.0)
    }

    struct TypographyTokens {
        static let display = Font.system(size: 28, weight: .bold)
        static let title = Font.system(size: 18, weight: .medium)
        static let body = Font.system(size: 14, weight: .regular) // Kotlin's FontWeight.Normal 对应 Swift 的 .regular
        static let label = Font.system(size: 12, weight: .medium)
    }

    struct Shapes {
        static let small = RoundedCornerShape(cornerRadius: 6)
        static let medium = RoundedCornerShape(cornerRadius: 10)
        static let large = RoundedCornerShape(cornerRadius: 16)
    }

    struct Spacing {
        static let sm: CGFloat = 6
        static let md: CGFloat = 10
        static let lg: CGFloat = 16
        static let xl: CGFloat = 24
        static let xxl: CGFloat = 36
    }

    struct ShadowSpec {
        let elevation: CGFloat
        let radius: CGFloat
        let dy: CGFloat
        let opacity: Double // Kotlin's Float 对应 Swift 的 Double
    }

    struct ElevationMapping {
        static let level1 = ShadowSpec(elevation: 2, radius: 4, dy: 2, opacity: 0.12)
        static let level2 = ShadowSpec(elevation: 6, radius: 8, dy: 4, opacity: 0.15)
        static let level3 = ShadowSpec(elevation: 10, radius: 12, dy: 6, opacity: 0.18)
    }
    
    // Material 3 默认尺寸/填充，用于精确匹配安卓行为
    struct MaterialDefaults {
        static let bottomAppBarHeight: CGFloat = 80
        static let fabSize: CGFloat = 56
        static let buttonHorizontalPadding: CGFloat = 16 // Material3 FilledButton 默认水平填充
        static let buttonVerticalPadding: CGFloat = 8 // Material3 FilledButton 默认垂直填充
        static let fabBottomMargin: CGFloat = 16 // FAB 距离底部边缘的默认间距
        static let fabTrailingMargin: CGFloat = 16 // FAB 距离右侧边缘的默认间距
        static let lazyColumnContentBottomPadding: CGFloat = 96 // Kotlin LazyColumn 的 contentPadding
    }
}

// 辅助结构体，用于封装圆角值
struct RoundedCornerShape {
    let cornerRadius: CGFloat
}

// MARK: - Data Model
// 列表数据模型，遵循 Identifiable 协议以便在 ForEach 中使用
struct LiveItem: Identifiable {
    let id: Int
    let title: String
    let viewers: Int
    let price: String
}

// MARK: - RootScreen View
struct RootScreen: View {
    let liveList: [LiveItem] = [
        LiveItem(id: 1, title: "Smartphone Flash Deal", viewers: 1234, price: "$699"),
        LiveItem(id: 2, title: "Gaming Chair Special", viewers: 880, price: "$249"),
        LiveItem(id: 3, title: "Headphones Clearance", viewers: 1640, price: "$99"),
        LiveItem(id: 4, title: "Mechanical Keyboard", viewers: 512, price: "$129")
    ]

    var body: some View {
        // ZStack 用于模拟 Scaffold 的层叠布局：内容、FAB 和底部导航栏
        ZStack {
            // 主要内容区域，对应 Scaffold 的 content
            VStack(alignment: .leading, spacing: AppTokens.Spacing.md) {
                Text("Live Shop Events")
                    .font(AppTokens.TypographyTokens.display)
                    .foregroundColor(AppTokens.Colors.primary)
                
                ScrollView {
                    LazyVStack(spacing: AppTokens.Spacing.md) {
                        ForEach(liveList) { item in
                            LiveItemCard(item: item)
                        }
                    }
                    // Kotlin LazyColumn 的 contentPadding 底部 96.dp，用于为 FAB 留出空间
                    .padding(.bottom, AppTokens.MaterialDefaults.lazyColumnContentBottomPadding)
                }
            }
            // 整个内容 Column 的外部填充，对应 Kotlin Modifier.padding(AppTokens.Spacing.lg)
            .padding(AppTokens.Spacing.lg)
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .background(
                LinearGradient( // 垂直渐变背景
                    gradient: Gradient(colors: [
                        AppTokens.Colors.background,
                        AppTokens.Colors.surfaceVariant,
                        AppTokens.Colors.background
                    ]),
                    startPoint: .top,
                    endPoint: .bottom
                )
            )
            .ignoresSafeArea(.all, edges: .all) // 使内容延伸至屏幕边缘，实现全屏效果
            
            // 浮动操作按钮 (Floating Action Button)
            VStack { // 使用 VStack + Spacer 将 FAB 推到 ZStack 的底部
                Spacer()
                HStack { // 使用 HStack + Spacer 将 FAB 推到 ZStack 的右侧
                    Spacer()
                    Button(action: {}) {
                        Text("+")
                            .font(AppTokens.TypographyTokens.title)
                            .foregroundColor(AppTokens.Colors.onSecondary)
                            .frame(width: AppTokens.MaterialDefaults.fabSize, height: AppTokens.MaterialDefaults.fabSize)
                            .background(AppTokens.Colors.secondary)
                            .clipShape(Circle()) // 圆形剪裁
                    }
                    // FAB 距离右侧和底部边缘的间距，底部间距需要考虑底部导航栏的高度
                    .padding(.trailing, AppTokens.MaterialDefaults.fabTrailingMargin)
                    .padding(.bottom, AppTokens.MaterialDefaults.fabBottomMargin + AppTokens.MaterialDefaults.bottomAppBarHeight)
                }
            }

            // 底部导航栏 (Bottom Bar)
            VStack { // 使用 VStack + Spacer 将底部导航栏推到 ZStack 的底部
                Spacer()
                BottomNavBar()
            }
        }
        .background(AppTokens.Colors.background) // Scaffold 的 containerColor
        .statusBarHidden(true) // 隐藏状态栏
    }
}

// MARK: - LiveItemCard View
struct LiveItemCard: View {
    let item: LiveItem

    var body: some View {
        VStack(alignment: .leading, spacing: AppTokens.Spacing.sm) {
            // 占位图片区域
            Rectangle()
                .fill(AppTokens.Colors.secondary)
                .frame(maxWidth: .infinity)
                .frame(height: 160)
                .clipShape(RoundedRectangle(cornerRadius: AppTokens.Shapes.medium.cornerRadius))

            Text(item.title)
                .font(AppTokens.TypographyTokens.title)
                .foregroundColor(AppTokens.Colors.onSurface)

            Text("\(item.viewers) viewers")
                .font(AppTokens.TypographyTokens.body)
                .foregroundColor(AppTokens.Colors.tertiary)

            HStack {
                Text(item.price)
                    .font(AppTokens.TypographyTokens.title)
                    .foregroundColor(AppTokens.Colors.primary)

                Spacer() // 将价格和按钮推开

                Button(action: {}) {
                    Text("Join Live")
                        .font(AppTokens.TypographyTokens.label)
                        .foregroundColor(AppTokens.Colors.onPrimary)
                        // 按钮内容填充，精确匹配 Material3 默认值
                        .padding(.vertical, AppTokens.MaterialDefaults.buttonVerticalPadding)
                        .padding(.horizontal, AppTokens.MaterialDefaults.buttonHorizontalPadding)
                        .background(AppTokens.Colors.primary)
                        .clipShape(RoundedRectangle(cornerRadius: AppTokens.Shapes.medium.cornerRadius))
                }
                .buttonStyle(PlainButtonStyle()) // 移除 SwiftUI 默认按钮样式
            }
        }
        .padding(AppTokens.Spacing.lg) // Card 内部填充
        .background(AppTokens.Colors.surface)
        .clipShape(RoundedRectangle(cornerRadius: AppTokens.Shapes.large.cornerRadius)) // Card 圆角
        .shadow(color: Color.black.opacity(AppTokens.ElevationMapping.level1.opacity), // Card 阴影
                radius: AppTokens.ElevationMapping.level1.radius,
                x: 0,
                y: AppTokens.ElevationMapping.level1.dy)
    }
}

// MARK: - BottomNavBar View
struct BottomNavBar: View {
    var body: some View {
        HStack(spacing: 0) { // 使用 spacing 0 配合 Spacer 实现 SpaceBetween 效果
            Spacer()
            Text("Home")
                .font(AppTokens.TypographyTokens.body)
                .foregroundColor(AppTokens.Colors.primary) // 主色
            Spacer()
            Text("Live")
                .font(AppTokens.TypographyTokens.body)
                .foregroundColor(AppTokens.Colors.onSurface)
            Spacer()
            Text("Cart")
                .font(AppTokens.TypographyTokens.body)
                .foregroundColor(AppTokens.Colors.onSurface)
            Spacer()
            Text("Profile")
                .font(AppTokens.TypographyTokens.body)
                .foregroundColor(AppTokens.Colors.onSurface)
            Spacer()
        }
        .frame(maxWidth: .infinity)
        .frame(height: AppTokens.MaterialDefaults.bottomAppBarHeight) // 底部导航栏高度
        .background(AppTokens.Colors.surface) // 背景色
        .shadow(color: Color.black.opacity(0.1), radius: 0.5, x: 0, y: -1) // 底部导航栏的顶部阴影
    }
}

// MARK: - App Entry Point
@main
struct LiveShopApp: App {
    var body: some Scene {
        WindowGroup {
            RootScreen()
        }
    }
}